car-bsp/Drivers/src/GpsDriver.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+ * Gps.cpp
+ *
+ *  Created on: June 16, 2023
+ *      Author: Raymond Bernardo, Noah Zinn, John Carr
+ */
+
+//Include Files
+#include "GpsDriver.hpp"
+#include "string.h"
+
+namespace SolarGators {
+namespace Drivers {
+
+Gps::Gps(UART_HandleTypeDef* uart_instance, SolarGators::DataModules::GPS* gps_data)
+{
+  gps_data_ = gps_data;
+  uart_instance_ = uart_instance;
+
+  uart_rx_event_ = osEventFlagsNew(NULL);
+  if (uart_rx_event_ == NULL)
+  {
+      Error_Handler();
+  }
+
+  rx_task_handle_ = osThreadNew((osThreadFunc_t)&Gps::HandleReceive, this, &rx_task_attributes_);
+  if (rx_task_handle_ == NULL)
+  {
+      Error_Handler();
+  }
+}
+
+Gps::~Gps()
+{ }
+
+void Gps::HandleReceive()
+{
+  while(1)
+  {
+    char gps_sentence[MAX_SENTENCE_LENGTH];
+    uint32_t flags = osEventFlagsWait(uart_rx_event_, (PING_FLAG | PONG_FLAG), osFlagsWaitAny, osWaitForever);
+    // These string copies need to happen fast
+    if(flags & PING_FLAG)
+    {
+      // Read from the ping buffer
+      strcpy(gps_sentence, reinterpret_cast<char*>(ping_));
+    }
+    else
+    {
+      // Read from the pong buffer
+      strcpy(gps_sentence, reinterpret_cast<char*>(pong_));
+    }
+    gps_data_->FromByteArray(reinterpret_cast<uint8_t*>(gps_sentence));
```

Comment:
This should be wrapped in a mutex so that `SendDataModule` doesn't get corrupted

```suggestion
    osMutexAcquire(gps_data_->mutex_id_, osWaitForever);
    gps_data_->FromByteArray(reinterpret_cast<uint8_t*>(gps_sentence));
    osMutexRelease(gps_data_->mutex_id_);      
```

We've already have a mutex ub `SendDataModule` :)
https://github.com/Solar-Gators/Sunrider-Firmware/blob/5c0d1dba7a54c896fc499b7a9621b5e7ddee5c60/car-bsp/Drivers/src/PitComms.cpp#L26